chain/events/observer.go

```diff
@@ -244,6 +244,41 @@ func (o *observer) headChange(ctx context.Context, rev, app []*types.TipSet) err
 	return nil
 }
 
+// ObserveAndBlock registers the observer and returns the current tipset along with an unlock function.
+//
+// This method guarantees that the observer will receive tipset updates starting from the returned tipset.
+// It blocks all tipset updates for all clients until the returned unlock function is called.
+//
+// The typical usage pattern is:
+// 1. Call ObserveAndBlock to register the observer
+// 2. Perform any necessary initialization using the returned current tipset
+// 3. Call the unlock function to start receiving updates
+//
+// Important notes:
+// - This method should only be called after the observer has been started
+// - The unlock function must be called to prevent blocking of tipset updates for all registered observers
+// - This method returns an error if the observer hasn't started yet
+//
+// Returns:
+// - *types.TipSet: The current tipset at the time of registration
+// - func(): An unlock function that must be called to start receiving updates
+// - error: An error if the observer hasn't started yet
+func (o *observer) ObserveAndBlock(obs TipSetObserver) (*types.TipSet, func(), error) {
```

Comment:
OK, so this lock only prevents new changes coming in, because we only lock around setting `o.head`, but you may call this and obtain the lock before, or while, an `Apply` or `Revert` is being processed. Is that going to cause any races with this pattern? Or should we retain the lock for the entirety of the `headChange` function?

Reply:
@rvagg 

Once the call to `ObserveAndBlock` returns, the caller is **_guaranteed_** two things:

    a) There will be NO more `Apply` and `Revert` calls till they release the lock using the handled returned here.
    b) Once they do release the lock, the first tipset update they will see will be for the tipset at **_OR_** immediately AFTER  the tipset returned by the call to ` ObserveAndBlock`.
    
**_Now:_**

1) If a caller obtains this lock before an `Apply ` or `Revert ` is processed -> great. The next `Apply` or `Revert` callback will start with the tipset returned here, no problem.
2) If a caller obtains this lock while an `Apply` or `Revert` is being processed -> they will definitely get the next update 

So we should be fine. **_However_**, looking at the mental gymnastics involved here, I am okay with just taking a lock on the whole thing and revisiting if it becomes a performance issue.

Reply:
@rvagg 

Once the call to ObserveAndBlock returns, the caller is **_guaranteed_** two things:
   - There will be NO more `Apply` and `Revert` calls till they release the lock using the handled returned here.
     
   -  Let's say the tipset returned by this function is `ts`. Once they do release the lock, the first tipset update they will see will be one of these even if an `Apply/Revert` is being processed **_while_** the caller holds the lock:
         - An apply from `ts` to `ts+1` OR
         - An apply from `ts-1` to `ts` OR
         - A revert from `ts` to `ts-1` OR 
         - A revert from `ts+1` to `ts` OR
  
  Given that our Indexing is idempotent, all of these are fine.

Reply:
@rvagg A cleaner/easier to reason about option here is for `obserer.start()` to: 

read the head from `ChainNotify -> return the head and an unlock handle -> block i.e. dont  start sending notifications yet -> caller calls the unlock handle -> start sending notifications.

This ensures that no `Apply`/`Revert` calls are made at all before re-conciliation is complete. Let me know what you think and I can make the change.

Reply:
I don't think I'm following because this still doesn't seem true to me:

> There will be NO more Apply and Revert calls till they release the lock using the handled returned here.

and in godoc:

> // It blocks all tipset updates for all clients until the returned unlock function is called.

Because of the lock release on lines 166, 205 and 231, so a call to `ObserveAndBlock` could grab the lock straight after it's released there, and an existing observer may have either an `Apply` or `Revert` called _after_ this function has returned. In fact, because there's 2 separate lock/unlock paths in `headChange` (3 if we had the case of both an apply and revert in the same change, but I don't think that's possible), there's an additional race I think: this function is called after the first lock.

But I think you're saying that this is all expected an ok, could be `ts` to `ts+1` _or_ `ts-1` to `ts`, etc. But then, the docs for this function doesn't seem quite right. Is the point simply to put the brakes on tipset processing for however long we might need.

> guarantees that the observer will receive tipset updates starting from the returned tipset

this is not true, it could be the returned tipset or the tipset+1.

it's the multiple locks within `headChange` that add uncertainty here, there's time before lock @ 166 and the time between lock @ 166 and @ 205/231, if you call this function and obtain the lock before or between then you'll have different outcomes.

Reply:
After discussing this offline with @rvagg , we have a couple of options here to reduce the cognitive overhead and make sure there's no spooky edge case because of the double locking mechanism used in `headChange`:

1) Keep around `ObserveAndBlock` but lock the entire `headChange` function. This ensures that the next update the client sees actually begins at `head`.

OR

2) Introduce an additional `StartAndBlock` API that returns the `head` that the the `Observer` gets from `ChainNotify` but then blocks on sending any updates whatsoever until the caller unblocks. This means that a client can reconcile only at start but that is our use case here anyways and makes things much easier to reason about and avoids the strict locking in option 1.

Also, fix the documentation of the `ObserveAndBlock` API.

Reply:
@rvagg I have implemented Option 1 here. It is the easiest to implement/reason about and we can revisit later if the locking is a problem at all but I doubt very much it will be given how the Observer really is only used at startup across the codebase.